### ISSUE TITLE
Customer Segmentation Project

### ISSUE THREAD

**Issue by akshata0710-hub** [akshata0710-hub/Customer-Segmentation-Project#1]
import pandas as pd
import numpy as np
import matplotlib.pyplot as  plt
import seaborn as sns
from sklearn.preprocessing import MinMaxScaler
from sklearn.metrics import silhouette_score
from sklearn.cluster import KMeans
from yellowbrick.cluster import KElbowVisualizer
data = pd.read_csv('/content/ecom_customer_data.csv')
data.head()
df=data.copy()
df.info()
df.describe()
#Data cleaning
#check the duplicates
df[df.duplicated()]
df['Gender']=df['Gender'].fillna(df['Gender'].mode()[0])
df.isna().sum().sum()
#Data visualization
df.Gender.value_counts()
sns.countplot(data=df,x='Gender')
plt.show()
plt.figure(figsize=(15,5))
plt.subplot(1,2,1)
sns.countplot(data=df,x='Orders')
#order count by each number
plt.subplot(1,2,2)
sns.countplot(data=df,x='Orders',hue='Gender')
plt.suptitle("overall Orders vs Gender wise Orders")
plt.show()
#Orders ans searches of each brands
cols=list(df.columns[2:])
def dist_list(lst):
    plt.figure(figsize=(30,30))
    for i,col in enumerate(lst,1):
        plt.subplot(6,6,i)
        sns.boxplot(data=df,x=df[col])
dist_list(cols)
#Heatmap
plt.figure(figsize=(20,15))
sns.heatmap(df.iloc[:,3:].corr())
plt.show()
df.iloc[:2,:].hist(figsize=(40,30))
plt.show()
new_df=df.copy()
new_df['Total Search']=new_df.iloc[:,3:].sum(axis=1)
new_df.sort_values('Total Search',ascending=False)
plt.figure(figsize=(13,8))
plt_data=new_df.sort_values('Total Search',ascending=False)[['Cust_ID','Gender','Total Search']].head(10)
sns.barplot(data=plt_data,x='Cust_ID',y='Total Search',hue='Gender',order=plt_data.sort_values('Total Search',ascending=False).Cust_ID)
plt.title("Top 10 Cust_ID based on Total Searches")
plt.show()
#scaling
x=df.iloc[:,2:].values
x
scale=MinMaxScaler()
features=scale.fit_transform(x)
features
#elbow method to get the optimal k value
inertia=[]
for i in range(1,16):
    k_means=KMeans(n_clusters=i)
    k_means=k_means.fit(features)
    inertia.append(k_means.inertia_)
#elbow graph
plt.figure(figsize=(20,7))
plt.subplot(1,2,1)
plt.plot(range(1,16),inertia,'bo-')
plt.xlabel('No of clusters'),plt.ylabel('Inertia')
#Kelbow visualizer
plt.subplot(1,2,2)
kmeans=KMeans()
visualize=KElbowVisualizer(kmeans,k=(1,16))
visualize.fit(features)
plt.suptitle("Elbow Graph and Elbow Visualizer")
visualize.poof()
plt.show()
#Silhouette score for each k value
silhouette_avg=[]
for i in range(2,16):
    #initialize kmeans cluster
    kmeans=KMeans(n_clusters=i)
    cluster_labels=kmeans.fit_predict(features)
    #Silhouette score
    silhouette_avg.append(silhouette_score(features,cluster_labels))
plt.figure(figsize=(10,7))
plt.plot(range(2,16),silhouette_avg,'bX-')
plt.xlabel('Values of K')
plt.ylabel('Silhouette score')
plt.title('Silhouette analysis for optimal K')
plt.show()
#Kmeans model here we will take k value as 3 as per Elbow method
model=KMeans(n_clusters=3)
model=model.fit(features)
y_km=model.predict(features)
centers=model.cluster_centers_
df['Cluster']=pd.DataFrame(y_km)
df.to_csv("Cluster_data",index=False)
#counts the no. of occurences of each value in cluster
df['Cluster'].value_counts()
sns.countplot(data=df,x='Cluster')
plt.show()
c_df=pd.read_csv('Cluster_data')
c_df.head()
c_df['Total Search']=c_df.iloc[:,3:38].sum(axis=1)
c1_0 = c_df.groupby(['Cluster', 'Gender'], as_index=False).sum().query('Cluster == 0')
c1_0
c1_1 = c_df.groupby(['Cluster', 'Gender'], as_index=False).sum().query('Cluster == 1')
c1_1
plt.figure(figsize=(15,6))
plt.subplot(1,2,1)
sns.countplot(data=c_df.query('Cluster==1'),x='Gender')
plt.title('Customers count')
plt.subplot(1,2,2)
sns.barplot(data=c1_1,x='Gender',y='Total Search')
plt.title('Total Searches by Gender')
plt.suptitle('No. of customer and their total searches in "Cluster 1"')
plt.show()
c1_2 = c_df.groupby(['Cluster', 'Gender'], as_index=False).sum().query('Cluster == 2')
c1_2
plt.figure(figsize=(15,6))
plt.subplot(1,2,1)
sns.countplot(data=c_df.query('Cluster==2'),x='Gender')
plt.title('Customers count')
plt.subplot(1,2,2)
sns.barplot(data=c1_1,x='Gender',y='Total Search')
plt.title('Total Searches by Gender')
plt.suptitle('No. of customer and their total searches in "Cluster 2"')
plt.show()
final_df = c_df.groupby(['Cluster']).sum().reset_index()
final_df
plt.figure(figsize=(15,6))
sns.countplot(data=c_df,x='Cluster',hue='Gender')
plt.title('Total Customer on each cluster')
plt.show()
plt.figure(figsize=(15,6))
plt.subplot(1,2,1)
sns.barplot(data=final_df,x='Cluster',y='Total Search')
plt.title('Total Searches by each group')
plt.subplot(1,2,2)
sns.barplot(data=final_df,x='Cluster',y='Orders')
plt.title('Past Orders by each group')
plt.suptitle('No. of customer searched the products and their past orders')
plt.show()
-----------------------------------------------------------------------------------------------------------------------------

[Customer Segmentation.pdf](https://github.com/user-attachments/files/15909202/Customer.Segmentation.pdf)

[ecom customer_data.xlsx](https://github.com/user-attachments/files/15909212/ecom.customer_data.xlsx)

https://github.com/akshata0710-hub/Customer-Segmentation-Project/assets/173335239/86e9f84e-bcd3-4287-96f3-e2b6deac822a
